docs/proposal-platforms.md

```diff
@@ -0,0 +1,259 @@
+    [[sugar]]
+        hosts = localhost
+        batch system = slurm
+    [[hpc]]
+        hosts = hpcl1, hpcl2
+        retrieve job logs = True
+        batch system = pbs
+    [[hpcl1-bg]]
+        hosts = hpcl1
+        retrieve job logs = True
+        batch system = background
+    [[hpcl2-bg]]
+        hosts = hpcl2
+        retrieve job logs = True
+        batch system = background
+[platform aliases]
+    [[hpc-bg]]
+        platforms = hpcl1-bg, hpcl2-bg
+```
+
+Note:
+
+* The default for `hosts` is the platform name and platform section headings can
+  be regular expressions to match multiple platform names. This allows us to
+  define many platforms in a single section and handles the case where you have
+  a large number of separate servers which are identical in terms of their
+  platform properties other than their host name.
+
+* The current logic for finding matching host settings is complicated and not
+  fully
+  [documented](https://cylc.github.io/doc/built-sphinx/appendices/site-user-config-ref.html#hosts-host)
+  (sections defining individual host names take precedence over pattern matches
+  regardless of where they come in the order). We propose a different logic for
+  finding matching platform settings:
+  * The pattern will be matched against the full name using
+    [re.fullmatch](https://docs.python.org/3/library/re.html#re.fullmatch)
+    (rather then `re.match` which is currently used for matching hosts).
+  * A hierarchy of platform match expressions from general to specific can be
+    used because config items are processed in the reverse order specified in the
+    file (i.e. bottom up). This is opposite to the method currently used for
+    matching hosts but is required because any additional platforms defined in
+    user config files get added to the end of the list. Currently it is
+    impossible to add hosts defined as regular expressions in the user config if
+    there is already a more generic regular expression defined in the global
+    config.
+  * There will be no special treatment of sections which define a single
+    platform. The first matching section found will be used.
+  * Note that the first reference to a section in the global or user config
+    determines its position in the list. You can repeat a section later in the
+    config and alter some settings but this doesn't change its position.
+
+* The names of all new and existing settings are subject to change until we
+  release cylc 8.0. Initially names will be chosen which match current usage
+  wherever possible. We intend to review all the config settings once we have
+  the key functionality in place.
+
+* The default platform is `localhost`. At cylc 7, any settings for `localhost`
+  act as defaults for all other hosts but we do **not** propose doing this for
+  platforms. The downside is more duplication if you need to override a
+  particular setting for every platform but we think this is clearer and safer.
```

Comment:
Perhaps we might want to think about the "inheritance" model. Would it be a good idea if we were able to do something along the lines of:

```ini

[*]
    # settings which apply to *all* platforms

[abc\d\d\d]
    # settings which apply to abc123

[abc*]
    # settings which also apply to abc123

```